### ISSUE TITLE
org.jboss.cdi.tck.tests.full.extensions.lifecycle.processBeanAttributes.specialization.VetoTest

### ISSUE THREAD

**Issue by jeanouii** [jakartaee/cdi-tck#440]
Per the current [TCK Process documentation](https://jakarta.ee/committees/specification/tckprocess/), please find bellow a challenge related to CDI TCK. Can someone please add the `challenge` label which is required by the TCK Process but not available to everyone to add.

### The relevant specification version and section number(s)

https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#bean_discovery_steps_full

### The coordinates of the challenged test(s)

Only following test in the class are related to the challenge 

* testSpecializedBeanAvailableAfterSpecializingBeanVetoed

### The exact TCK version

CDI TCK version 4.0.7

### The implementation being tested, including name and company

Apache OpenWebBeans 4.0.0-SNAPSHOT from the Apache Software Foundation

### A full description of why the test is invalid and what the correct behavior is believed to be

The test has 3 types discovered: Alpha, Bravo and Charlie. Bravo specializes Alpha and Charlie specializes Bravo. 
The specification says that for each discovered type, the container has to trigger events such as ProcessBeanAttributes.
In this test, we have 2 CDI extensions, one to veto Charlie when PBA is triggered and another one to check if PBA has been triggered for Alpha, Bravo and Charlie.

The test expects either an order to process the discovered types in the loop, Charlie, Bravo, Alpha, or it implies after vetoing Charlie another loop needs to happen to process again Bravo which is not specialized anymore.

An implementation may process Bravo first and not come process it again after Charlie is vetoed.

**Comment by manovotn** [comment]
>  The specification says that for each discovered type, the container has to trigger events such as ProcessBeanAttributes.

The specification says that for each **_enabled_** bean, the container has to trigger events... (at least for PBA and few others it is stated like this).
The specification further described what an enabled bean is [in this section](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enablement_full), specifically saying:

> A bean is said to be enabled if it is not specialized by any other enabled bean, as defined in [Specialization](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#specialization)

For any onlookers, there is a detailed discussion about this topic on the mailing list, starting with [this post](https://www.eclipse.org/lists/cdi-dev/msg00698.html).


Personally, I do not consider this challenge to be valid.

**Comment by manovotn** [comment]
> So the crucial question is: does PBA.veto() on a specialization cause enablement change on the specialized superclass?

So assuming for a while it would *not* change the enablement - that is, `Bravo` would still be considered to be a *disabled* bean after you `PBA#veto()` `Charlie`.
The spec has some requirements on enabled beans that `Bravo` now wouldn't meet (this isn't complete list of occurrences, just highlights):

- Any observer on `Bravo` wouldn't work as those need to be on enabled beans, [spec link](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#observer_resolution)
- `Bravo` is *not* subject to typesafe resolution validation and could in theory have ambig. or unsatisfied injection points without CDI telling you, [spec link](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#unsatisfied_and_ambig_dependencies)
- Producers would also be skipped since container only searches enabled beans for them, [spec link](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#bean_discovery_steps)
- There doesn't need to be an instance of `Bean<Bravo>` present in the CDI container because the spec literally says that there must be an instance of `Bean` for each *enabled* bean, [spec link](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#bean)

In other words, if `Bravo` remained disabled, you wouldn't be able to use it in typesafe resolution and therefore any injection point of type `Alpha` or `Bravo` would then be unsatisfied.
Or am I misinterpreting it @Ladicek?

**Comment by Ladicek** [comment]
I don't know. I understand what does it mean to veto a type (during type discovery at latest), but the spec is not very talkative about the meaning of vetoing a _bean_ (during bean discovery). That will always end up in some form of perceived inconsistency.

I overall agree that for the spec to remain as consistent as possible, vetoing a specialization _should_ result in changed enablement of the specialized superclass, but inconsistency is Portable Extension's middle name, so... :shrug:

**Comment by jeanouii** [comment]
Here are the exact words from the spec

> For every type in the set of discovered types (as defined in [Type discovery in CDI Full](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#type_discovery_steps_full)), the container must:
>
>  -  inspect the type metadata to determine if it is a bean, and then
>
>  -  detect definition errors by validating the class and its metadata, and then
>
>  -  if the class is a managed bean, fire an event of type ProcessInjectionPoint for each injection point in the class, as defined in [ProcessInjectionPoint event](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#process_injection_point), and then
>
>  -  if the class is a managed bean, fire an event of type ProcessInjectionTarget, as defined in [ProcessInjectionTarget event](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#process_injection_target), and then
>
>  -  determine which alternatives, interceptors and decorators are enabled, according to the rules defined in [Enabled and disabled beans](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enablement), [Interceptor enablement and ordering in CDI Full](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enabled_interceptors) and [Decorator enablement and ordering](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enabled_decorators), and then
>
>  -  if the class is an enabled bean, interceptor or decorator, fire an event of type ProcessBeanAttributes, as defined in [ProcessBeanAttributes event](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#process_bean_attributes), and then
>
>  -  if the class is an enabled bean, interceptor or decorator and if ProcessBeanAttributes.veto() wasn’t called in previous step, fire an event which is a subtype of ProcessBean, as defined in [ProcessBean event](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#process_bean).

"For every type in the set of discovered types" --> Means there is a loop. No order specified. Does not say we have to reprocess discovered types under certain unspecified conditions. Note the accurate use of "discovered types" because there is no Bean quite yet.

"determine which alternatives, interceptors and decorators are enabled, according to the rules defined in [Enabled and disabled beans]" --> BTW, it does not speak about enablement of managed beans, only alternatives, interceptors and decorators.

"if the class is an enabled bean, interceptor or decorator and if ProcessBeanAttributes.veto() wasn’t called in previous step, fire an event which is a subtype of ProcessBean" --> does not say we have to restart again for beans that changed from not enabled to enabled. The test expects this behavior hence the challenge.

The spec is very very light in the way specialization have to be handled. The validation happens at the end. Fine. But I agree with Martej, this creates a big mess for a small gain we could solve with extensions. Ideally, it would be great to have something like "Adding @Specializes on a bean is equivent to having this extension (with an example)."

A friend of mine from the university would say "From my (academic / scientific) point of view: If other documents are needed to explain something within the actual spec, the spec is not clear." Either a specification is self contained/explained or it must explicitely mentioned external resources required with the right version when needed (like interceptors), or the spec isn't clear. 

I feel like the challenges are taken maybe personnaly or as a problem but they aren't. Challenges are very positive for a specification. It means the specification is reviewed and read by others, it means there are more implementations, and it means we can reach a better portability.

> inconsistency is Portable Extension's middle name

Ah ah, good one @Ladicek 
Made me smile :-)

**Comment by Ladicek** [comment]
From my perspective, the challenges you've been filing have all been very good. I don't think anyone is taking them personally. But note that what you're challenging is the TCK, not the specification. Therefore, as you can see here, we need to discuss whether the specification supports the TCK test or not -- that's what you see here. I believe it is better to conduct that discussion in the open.

I also believe that the specification does not have to "say we have to restart again" -- if "restarting again" is a necessary requirement to satisfy the specification, then you have to do it even if the specification doesn't say so explicitly. Implicit requirements are as valid as explicit. The discussion here is really about this: does the implicit requirement exist? I honestly don't know.

**Comment by jeanouii** [comment]
Thanks for the kind words. I don't want to be a pain. And I'd like to really also help on the CDI part whenever I'm done with the remaining issues I'm looking at.

**Comment by manovotn** [comment]
> I feel like the challenges are taken maybe personnaly or as a problem but they aren't. Challenges are very positive for a specification. It means the specification is reviewed and read by others, it means there are more implementations, and it means we can reach a better portability.

Sorry if my comments are making you feel that way - it's definitely not meant like that.
On the contrary, your questions and challenges are very clear and meaningful and I am trying to provide my best feedback on it, although it may come out rough at times.

> I also believe that the specification does not have to "say we have to restart again" -- if "restarting again" is a necessary requirement to satisfy the specification, then you have to do it even if the specification doesn't say so explicitly. Implicit requirements are as valid as explicit.

This is well said.
I'd also add that TCKs, while they need to have the ability to be challenged, should also be viewed as a "source of truth" to a certain degree. After all, TCKs may test these implicit requirements and OFC for such a test you then won't find an exact word-by-word backup in the specification text.
For instance in this particular case, I admit I am fairly amazed that we are talking a 9+ years old test (several major versions) that hasn't been change since and hadn't been added any later than specialization was. And yet what we are seeing is that other implementations settled with excluding this test ever since it's inception instead of passing the TCKs or actively trying to suggest changes. Seeing this just makes it a lot harder to believe the portability talk is all I am saying :-)

**Comment by Ladicek** [comment]
I'd like to keep this progressing.

First, I'd like to address these 2 paragraphs:

> determine which alternatives, interceptors and decorators are enabled, according to the rules defined in [Enabled and disabled beans](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enablement), [Interceptor enablement and ordering in CDI Full](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enabled_interceptors) and [Decorator enablement and ordering](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enabled_decorators), and then
>
> if the class is an enabled bean, interceptor or decorator, fire an event of type ProcessBeanAttributes, as defined in [ProcessBeanAttributes event](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#process_bean_attributes), and then

Clearly the 1st item here only talks about determining enablement of alternatives, interceptors and decorators. It might seem that the CDI container doesn't have to determine enablement of specializations in this step, because the specification doesn't say so.

However, the 2nd item here clearly says "if the class is an enabled bean, interceptor or decorator". This implies that "all" enablement needs to be determined, otherwise how can you say whether a class is an enabled bean or not?

To me, this means that the 1st item should be clarified [1], but isn't itself enough to determine whether this challenge is valid or not.

To be honest, I can't find anything in the specification that says or implies that vetoing a specialization causes enablement change on the superclass. I believe it is the logically most consistent way, because not causing enablement change on the superclass leaves it disabled, which is against the spirit of enablement and vetoing as I understand it. But, as mentioned, I don't see how that is required anywhere ("common sense" and "the TCK has been like that since forever" are both good to bring up, but don't really count in the end). The specification should either require this to happen (explicitly or implicitly), or should say that vetoing a specialization during PBA results in non-portable behavior.

[1] I'd suggest the following rewording:

> determine which beans (including alternatives, interceptors, decorators and specializations) are enabled, according to the rules defined in [Enabled and disabled beans in CDI Full](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enablement_full), [Interceptor enablement and ordering in CDI Full](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enabled_interceptors) and [Decorator enablement and ordering](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#enabled_decorators), and then

**Comment by manovotn** [comment]
> To me, this means that the 1st item should be clarified [1]

Sure, we can do that to have it explicitly stated in the spec but that changes nothing WRT this issue.

> ... because not causing enablement change on the superclass leaves it disabled, ...

Right, which leads to (among other things) what I pointed in the earlier comment(https://github.com/jakartaee/cdi-tck/issues/440#issuecomment-1453526145) and that wouldn't really be a desirable behavior - it'd basically mean that disabling a specialized bean most likely leads to a deployment error because of unsatisfied dependency.

> The specification should either require this to happen (explicitly or implicitly), or should say that vetoing a specialization during PBA results in non-portable behavior.

I agree it's not explicit in the spec, OTOH I don't see what else should the spec say to make it implicit - that's IMO already there, along with a TCK explicitly asserting this implicit requirement :shrug:
We can propose a change to spec that explicitly states that there is re-enablement if this happens but for that we'd first need to agree that the current behavior is expected and we need a clarification and that's not the case so far :)

I was thinking of the non-portable behavior as well but that solves very little because users who observe `PBA<? extends Foo>` don't have any notion of whether that `Bean` is specializing other bean and may end up unknowingly relying on non-portable behavior. Not to mention it obviously doesn't make the spec any better or more reliable.

**Comment by Ladicek** [comment]
Vetoing any bean in general may lead to a deployment problem due to a suddenly-unsatisfied dependency, so I'm afraid this argument doesn't cut it. I mean, you could argue that the ability to veto a bean should be removed, because it can cause inconsistencies :-)

What the specification should say, in my opinion, is _what does it mean to veto a bean_ (and an observer method, but that's unrelated to the present issue). This is extremely underspecified. I can understand what does it mean to veto a type, though it isn't exactly well specified either, but vetoing a bean is super unclear.

If vetoing means "from now on, the container should behave as if the vetoed thing never existed", that's pretty easy to interpret for types (the set of discovered types at the end of type discovery will not include the type, and bean discovery obviously proceeds on the modified set), but it leaves open questions when it comes to beans. The order in which the container processes discovered types during bean discovery is nondeterministic, so when __strictly following the definition above__, vetoing a specialization _before_ its superclass is processed leads to registering the superclass as a bean, but vetoing a specialization _after_ its superclass is processed leads to _not_ registering the superclass as a bean. Clearly this definition (which is really just my personal interpretation of the specification) is not good enough.

We need to define vetoing properly.

**Comment by manovotn** [comment]
@jeanouii @Ladicek I've created a proposal for clarification, see https://github.com/jakartaee/cdi/pull/658

**Comment by jeanouii** [comment]
Sorry guys, I have been working on something else and could not get into this issue earlier.

I agree with @Ladicek "it's been this way for 10 years isn't a valid argument"

> I'd also add that TCKs, while they need to have the ability to be challenged, should also be viewed as a "source of truth" to a certain degree. After all, TCKs may test these implicit requirements and OFC for such a test you then won't find an exact word-by-word backup in the specification text.

This is not correct. The single and only source of truth is the specification. TCK are not meant to complement the specification. 

The 3 of us have a slightly different reading looks like. Most of the tests come from Weld (this one included) which is one view of the specification from an implementor but does not necessarly represent what people meant/agreed to write on the spec. And this is why it creates problems. The is not the way it's suppose to work. Spec team should work together to create the specification, identify the assertions and create the tests accordingly. At least this is how we used to work with the JCP. I remember Linda explaining that somewhere during a conference. I'm not pointing fingers or blaming anyone. I think it's great to have a lot of tests available, but some of them coming from Weld are testing Weld not what the spec requires.

CDI lite and core profile are now part of Jakarta EE. It may bring us some more players and who knows another CDI implementation or newer usages. I don't give it a single chance based on the current wording for someone to come up with what Weld implemented. And this is the point of me creating this challenge

Second thing is how we can improve the specification. I read the proposal from @Ladicek and I don't mind but as we all agree specialization is a big mess (for a low value). Isn't it the time to address the specialization as a whole and for the entire spec as opposed to adding a small clarification in one place at the risk of creating other issues? Maybe this is enough and then all good. Maybe not. At this moment, I can't say to be honest. Maybe something to discuss during a call so we can have more fluent exchanges. We can capture that into the CDI Issue @Ladicek created after discussion. We also have to collect all the related tests/assertions and make sure they fit

**Comment by Ladicek** [comment]
Just agreed on the CDI call, we agreed that this challenge should be accepted and the test should be excluded.

In the next version of the specification, the behavior in question should be clarified and the test either re-enabled or reworked.
